### PR TITLE
Forward links propagation

### LINKED ISSUE #1793: missing forward links when roster changes
Seen on the production cothority, where the 3 original nodes have a forward link for 0->16, but nodes added into the roster after block 16 are missing the 0->16 link. Thus from the point of view of accordion.employees.org:7770, the shortest path from 0->21 is 4 blocks long. But form the point of view of fairywren.ch:7770, the path is 7 nodes:

In the following, `scmgr s u` is a new feature to call GetUpdateChainLevel on targeted nodes. 3.toml is accordion.employees.org:7770, 4.toml is fairywren.ch:7770.

```bash
$ scmgr s u 3.toml -from 9cc36071ccb902a1de7e0d21a2c176d73894b1cf88ae4cc2ba4c95cd76f474f3
index 0, hash 9cc36071ccb902a1de7e0d21a2c176d73894b1cf88ae4cc2ba4c95cd76f474f3
index 16, hash c287771fd714b79a1b4ad60ef5c2779f580215d0ea9dd14e37eaf23e5f6e93ad
index 20, hash 1065f9544f65750719ba6a8c81a85d8012e8e9371478607110564d2171da55d9
index 21, hash fbf0896b9dda00b378fe710f77457ad1432af218dec3ac1433ca05daa84996ff
$ scmgr s u 4.toml -from 9cc36071ccb902a1de7e0d21a2c176d73894b1cf88ae4cc2ba4c95cd76f474f3
index 0, hash 9cc36071ccb902a1de7e0d21a2c176d73894b1cf88ae4cc2ba4c95cd76f474f3
index 4, hash 418c1c9748752db450fdf851ae93ae71e0ba23d35f61f1d9eb350e48a7cc8b25
index 8, hash 50644255b43a4e425b5534bf60197cdbc31f7e1501be77e913893820fb667796
index 12, hash 291d7445836a09844594bd663558539e6f1636be2bac0ab5c80f52dda8f4ab0c
index 16, hash c287771fd714b79a1b4ad60ef5c2779f580215d0ea9dd14e37eaf23e5f6e93ad
index 20, hash 1065f9544f65750719ba6a8c81a85d8012e8e9371478607110564d2171da55d9
index 21, hash fbf0896b9dda00b378fe710f77457ad1432af218dec3ac1433ca05daa84996ff
```

Todo:
- [ ] write a unit test to trigger this situation
- [ ] with the unit test, investigate how the chain behaves after it, i.e. will higher level forward links be correctly added?
- [ ] probably fix the call to `s.startPropagation(s.propagateForwardLink, from.Roster, &PropagateForwardLink{fl, fs.TargetHeight})` to use the new Roster instead of from.Roster
- [ ] decide if/when/how to catch up missed forward links

### PR DESCRIPTION
**Pull request by Gilthoniel**
This PR fixes the propagation of the forward links by using the most recent roster to propagate to active conodes. It also changes the way forward links are updated to verify that the heights correspond to the index.

Also: clarify the situation with a base height of 1

Fixes #1793

### CODE REVIEW

**Comment by Gilthoniel** [comment]
Merging this PR will conflict with #1827. A rebase will be necessary to update the forward link verify to use the signature scheme.

File: skipchain/long_test.go [hunk 1]
```diff
@@ -113,3 +114,35 @@ func TestFail(t *testing.T) {
 		local.CloseAll()
 	}
 }
+
+func TestForwardLinks(t *testing.T) {
```
Comment (jeffallen) [inline skipchain/long_test.go]: Can we have a comment explaingin what's special about this test, what the goal is, etc?

File: skipchain/struct.go [hunk 2]
```diff
@@ -706,23 +711,41 @@ func (db *SkipBlockDB) StoreBlocks(blocks []*SkipBlock) ([]SkipBlockID, error) {
...
+						}
+
+						target, err := db.getFromTx(tx, fl.To)
+						if err != nil {
+							return err
+						}
+						if target != nil {
+							diff := math.Log(float64(target.Index - sbOld.Index))
+							base := math.Log(float64(sbOld.BaseHeight))
+							// Only check the target height if it exists because the block might
```
Comment (jeffallen) [inline skipchain/struct.go]: Comment before the if != nil please.

File: skipchain/skipchain.go [hunk 3]
```diff
@@ -1235,10 +1235,11 @@ func (s *Service) forwardLink(req *network.Envelope) error {
 			return err
 		}
 
-		// BackLinks are already stored in the new block so only
-		// the roster from the previous block needs to add the
-		// forward link
-		return s.startPropagation(s.propagateForwardLink, from.Roster, &PropagateForwardLink{fl, fs.TargetHeight})
+		// Forward-links are sent to the new roster so active conodes get the update. If a conode
```
[the remark this answers is not in the document]
Comment (Gilthoniel) [reply skipchain/skipchain.go]: That's right, I'll add a nil check.

**Pull request merged**